node/network/availability-distribution/src/requester.rs

```diff
@@ -0,0 +1,226 @@
+//! availability.
+
+use std::collections::{
+	hash_map::{Entry, HashMap},
+	hash_set::HashSet,
+};
+use std::iter::IntoIterator;
+use std::pin::Pin;
+use std::sync::Arc;
+
+use futures::{
+	channel::mpsc,
+	task::{Context, Poll},
+	Stream,
+};
+
+use sp_keystore::SyncCryptoStorePtr;
+
+use polkadot_node_subsystem_util::request_availability_cores_ctx;
+use polkadot_primitives::v1::{CandidateHash, CoreState, Hash, OccupiedCore};
+use polkadot_subsystem::{
+	messages::AllMessages, ActiveLeavesUpdate, jaeger, SubsystemContext,
+};
+
+use super::{error::recv_runtime, session_cache::SessionCache, Result, LOG_TARGET};
+
+/// A task fetching a particular chunk.
+mod fetch_task;
+use fetch_task::{FetchTask, FetchTaskConfig, FromFetchTask};
+
+/// Requester takes care of requesting erasure chunks from backing groups and stores them in the
+/// av store.
+///
+/// It implements a stream that needs to be advanced for it making progress.
+pub struct Requester {
+	/// Candidates we need to fetch our chunk for.
+	///
+	/// We keep those around as long as a candidate is pending availability on some leaf, so we
+	/// won't fetch chunks multiple times.
+	fetches: HashMap<CandidateHash, FetchTask>,
+
+	/// Localized information about sessions we are currently interested in.
+	session_cache: SessionCache,
+
+	/// Sender to be cloned for `FetchTask`s.
+	tx: mpsc::Sender<FromFetchTask>,
+
+	/// Receive messages from `FetchTask`.
+	rx: mpsc::Receiver<FromFetchTask>,
+}
+
+impl Requester {
+	/// Create a new `Requester`.
+	///
+	/// You must feed it with `ActiveLeavesUpdate` via `update_fetching_heads` and make it progress
+	/// by advancing the stream.
+	pub fn new(keystore: SyncCryptoStorePtr) -> Self {
+		// All we do is forwarding messages, no need to make this big.
+		let (tx, rx) = mpsc::channel(1);
+		Requester {
+			fetches: HashMap::new(),
+			session_cache: SessionCache::new(keystore),
+			tx,
+			rx,
+		}
+	}
+	/// Update heads that need availability distribution.
+	///
+	/// For all active heads we will be fetching our chunks for availabilty distribution.
+	pub async fn update_fetching_heads<Context>(
+		&mut self,
+		ctx: &mut Context,
+		update: ActiveLeavesUpdate,
+	) -> Result<()>
+	where
+		Context: SubsystemContext,
+	{
+		let ActiveLeavesUpdate {
+			activated,
+			deactivated,
+		} = update;
+		// Order important! We need to handle activated, prior to deactivated, otherwise we might
+		// cancel still needed jobs.
+		self.start_requesting_chunks(ctx, activated.into_iter())
+			.await?;
+		self.stop_requesting_chunks(deactivated.into_iter());
+		Ok(())
+	}
+
+	/// Start requesting chunks for newly imported heads.
+	async fn start_requesting_chunks<Context>(
+		&mut self,
+		ctx: &mut Context,
+		new_heads: impl Iterator<Item = (Hash, Arc<jaeger::Span>)>,
+	) -> Result<()>
+	where
+		Context: SubsystemContext,
+	{
+		for (leaf, _) in new_heads {
+			let cores = query_occupied_cores(ctx, leaf).await?;
```

Comment:
Previous behavior was also to look at `k` ancestors in same session so that even once 2/3+ of validators have their piece, there is still a little bit of time for the remaining 1/3 to fetch theirs. Even `k=1` is useful. As long as the blocks are all in the same session.

Reply:
Not sure I can follow. The requesting validator requests, what it sees pending availability in the blocks it is aware of.

Reply:
Let's say `ActiveLeaves = {X}`. In X pending availability is `{A}`. While `X in ActiveLeaves`, all validators will work to fetch their chunk.

However, it only takes 2/3 of validators to achieve availability on a candidate. Let's say in the next block Y, a descendant of X, there is nothing pending availability. Then if there are 1/3 of validators who are slow, they'll just stop requesting their chunk entirely. 

What we did previously was combine pending availability sets of a block with `k` of their ancestors. So even though the candidate is available in Y and thus isn't pending availability, we'd still consider everything pending availability in X as relevant for Y. 

This buys something like `k*slot_duration` extra time for validators experiencing network asynchrony to fetch their chunks and achieves better overall distribution.

Reply:
So slow means, ActiveLeaveUpdate skips a block (or two)? So we would not see the block where a candidate was pending availability?

Reply:
Yes, either we don't see it, or we see it only shortly before the next block. At which point 2/3 already have their piece.

Reply:
Not a blocker but shouldn't be hard to do in a follow-up as there already exists code for this. Filed https://github.com/paritytech/polkadot/issues/2513